ws/src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java

```diff
@@ -112,12 +117,15 @@ private void encodeFrame(WsURLConnectionImpl connection, Frame frame) throws IOE
             mask = 1 + connection.getRandom().nextInt(Integer.MAX_VALUE);
         }
 
-        out.write(buf.get(offset));
-        encodePayloadLength(out, payloadLength);
-        encodeMaskAndPayload(out, buf, payloadOffset, payloadLength, mask);
+        frameBuffer[frameBufferOffset++] = buf.get(offset);
+        encodePayloadLength(payloadLength);
```

Comment:
Perhaps we can block copy the pre-encoded metadata at the head of the frame (instead of re-computing it), apply the mask bit and the mask bytes to the frame buffer metadata directly, and then block mask & transfer the payload (4 bytes at a time instead of 1 byte at a time).